---
title: "DifferentialGeneExpressionAnalysis"
author: "Mia Anscheit"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/Users/matanatmammadli/Desktop/Project')

if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

#BiocManager::install("TCGAbiolinks")
#BiocManager::install("DESeq2")

library(TCGAbiolinks)
library(DESeq2)
library(DT)
library(dplyr)
```

## Load data 

```{r data}
# LARYNX
HNSC_meta <- read.delim('/Users/matanatmammadli/Desktop/Project/HNSC_exposure.tsv')
HNSC_cohort <- read.delim('/Users/matanatmammadli/Desktop/Project/HNSC_aliquot.tsv', sep = '\t', header = TRUE, fill = TRUE)

# LUNG
LUSC_meta <- read.delim('/Users/matanatmammadli/Desktop/Project/LUSC_exposure.tsv')
LUSC_cohort <- read.delim('/Users/matanatmammadli/Desktop/Project/LUSC_aliquot.tsv', sep = '\t', header = TRUE, fill = TRUE)
```

## Data cleaning (remove missing data)

```{r cleaning}
# LARYNX
HNSC_missing <- which(HNSC_meta$cigarettes_per_day == "'--")
HNSC_metadata <- HNSC_meta[- HNSC_missing,]
HNSC_case_ids <- HNSC_metadata$case_id
HNSC_cohortdata <- HNSC_cohort %>%
  filter(case_id %in% HNSC_case_ids)

# LUNG
LUSC_missing <- which(LUSC_meta$cigarettes_per_day == "'--")
LUSC_metadata <- LUSC_meta[- LUSC_missing,]
LUSC_case_ids <- LUSC_metadata$case_id
LUSC_cohortdata <- LUSC_cohort %>%
  filter(case_id %in% LUSC_case_ids)
```

## Download counts

```{r download}
# LARYNX
HNSC_submitter_ids <- HNSC_cohortdata$aliquot_submitter_id

HNSC_query <- GDCquery(
  project = "TCGA-HNSC", 
  data.category = "Transcriptome Profiling", 
  data.type = "Gene Expression Quantification", 
  workflow.type = "STAR - Counts",
  barcode =  HNSC_submitter_ids
)

# remove samples from non-tumor tissue
HNSC_tumor <- HNSC_query[[1]][[1]][HNSC_query[[1]][[1]]$sample_type == "Primary Tumor", ] 
HNSC_tumor_cases <- HNSC_tumor$cases

HNSC_tumor_query <- GDCquery(
  project = "TCGA-HNSC", 
  data.category = "Transcriptome Profiling", 
  data.type = "Gene Expression Quantification", 
  workflow.type = "STAR - Counts",
  barcode =  HNSC_tumor_cases
)

GDCdownload(HNSC_tumor_query)
HNSC <- GDCprepare(query = HNSC_tumor_query)


# LUNG
LUSC_submitter_ids <- LUSC_cohortdata$aliquot_submitter_id

LUSC_query <- GDCquery(
  project = "TCGA-LUSC", 
  data.category = "Transcriptome Profiling", 
  data.type = "Gene Expression Quantification", 
  workflow.type = "STAR - Counts",
  barcode =  LUSC_submitter_ids
)

# remove samples from non-tumor tissue
LUSC_tumor <- LUSC_query[[1]][[1]][LUSC_query[[1]][[1]]$sample_type == "Primary Tumor", ]
LUSC_tumor_cases <- LUSC_tumor$cases

LUSC_tumor_query <- GDCquery(
  project = "TCGA-LUSC", 
  data.category = "Transcriptome Profiling", 
  data.type = "Gene Expression Quantification", 
  workflow.type = "STAR - Counts",
  barcode =  LUSC_tumor_cases
)

GDCdownload(LUSC_tumor_query)
LUSC <- GDCprepare(query = LUSC_tumor_query)
```

## create count data table

```{r count data}
# LARYNX
HNSC_files <- list.files(path="GDCdata/TCGA-HNSC/Transcriptome_Profiling/Gene_Expression_Quantification/", recursive=T)

# initialization of count data matrix
HNSC_path <- file.path('GDCdata/TCGA-HNSC/Transcriptome_Profiling/Gene_Expression_Quantification/', HNSC_files[1])
HNSC_data <- read.delim(file=HNSC_path, skip=1, sep="", header = T, fill = TRUE) # load first file
HNSC_data <- HNSC_data[5:nrow(HNSC_data),] # remove total number information
HNSC_countdata <- HNSC_data[, "unstranded", drop = FALSE] # take unstranded columns as count data 
rownames(HNSC_countdata) <- paste0(HNSC_data$gene_id) # add gene ids as row names
colnames(HNSC_countdata) <- strsplit(basename(HNSC_files[1]), "\\.")[[1]][1] # add sample id

# create count data matrix
for (i in 2:length(HNSC_files)){
  HNSC_path <- file.path('GDCdata/TCGA-HNSC/Transcriptome_Profiling/Gene_Expression_Quantification/', HNSC_files[i])
  HNSC_data <- read.delim(HNSC_path, skip=1, sep="", header = T, fill = TRUE)
  HNSC_data <- HNSC_data[5:nrow(HNSC_data),]
  HNSC_count <- HNSC_data[, "unstranded", drop = FALSE]
  colnames(HNSC_count) <- strsplit(basename(HNSC_files[i]), "\\.")[[1]][1]
  HNSC_countdata <- cbind(HNSC_countdata, HNSC_count)
}


# LUNG
LUSC_files <- list.files(path="GDCdata/TCGA-LUSC/Transcriptome_Profiling/Gene_Expression_Quantification/", recursive=T)

# initialization of count data matrix
LUSC_path <- file.path('GDCdata/TCGA-LUSC/Transcriptome_Profiling/Gene_Expression_Quantification/', LUSC_files[1])
LUSC_data <- read.delim(file=LUSC_path, skip=1, sep="", header = T, fill = TRUE) # load first file
LUSC_data <- LUSC_data[5:nrow(LUSC_data),] # remove total number information
LUSC_countdata <- LUSC_data[, "unstranded", drop = FALSE] # take unstranded columns as count data 
rownames(LUSC_countdata) <- paste0(LUSC_data$gene_id) # add gene ids as row names
colnames(LUSC_countdata) <- strsplit(basename(LUSC_files[1]), "\\.")[[1]][1] # add sample id

# create count data matrix
for (i in 2:length(LUSC_files)){
  LUSC_path <- file.path('GDCdata/TCGA-LUSC/Transcriptome_Profiling/Gene_Expression_Quantification/', LUSC_files[i])
  LUSC_data <- read.delim(LUSC_path, skip=1, sep="", header = T, fill = TRUE)
  LUSC_data <- LUSC_data[5:nrow(LUSC_data),]
  LUSC_count <- LUSC_data[, "unstranded", drop = FALSE]
  colnames(LUSC_count) <- strsplit(basename(LUSC_files[i]), "\\.")[[1]][1]
  LUSC_countdata <- cbind(LUSC_countdata, LUSC_count)
}
```

## get coldata

```{r coldata}
# LARYNX
HNSC_coldata <- HNSC_tumor_query[[1]][[1]]

# set correct ids for rownames (so the ids in count and col data are the same)
HNSC_rown <- c()
for (i in 1:nrow(HNSC_coldata)) {
  r <- strsplit(HNSC_coldata$analysis_submitter_id[i], "_star__counts")[[1]]
  HNSC_rown <- c(HNSC_rown, r)
}
rownames(HNSC_coldata) <- HNSC_rown

# sort meta data (so order inis the same like col in count data
HNSC_sort_order_indices <- match(rownames(HNSC_coldata), colnames(HNSC_countdata))
HNSC_coldata <- HNSC_coldata[order(HNSC_sort_order_indices), ]


# LUNG
LUSC_coldata <- LUSC_tumor_query[[1]][[1]]

# new rownames
LUSC_rown <- c()
for (i in 1:nrow(LUSC_coldata)) {
  r <- strsplit(LUSC_coldata$analysis_submitter_id[i], "_star__counts")[[1]]
  LUSC_rown <- c(LUSC_rown, r)
}
rownames(LUSC_coldata) <- LUSC_rown

# sort meta data, so order is the same like col in count data
LUSC_sort_order_indices <- match(rownames(LUSC_coldata), colnames(LUSC_countdata))
LUSC_coldata <- LUSC_coldata[order(LUSC_sort_order_indices), ]
```

## combine larynx and lung for DESeq2 (& change - to _ and factorize row)

```{r combine}
countdata <- cbind(HNSC_countdata, LUSC_countdata)
coldata <- rbind(HNSC_coldata, LUSC_coldata)

coldata$project <- gsub("-", "_", coldata$project)
coldata$project <- factor(coldata$project)
```

## create dds object and filter out low counts

```{r dds}
dds <- DESeqDataSetFromMatrix(countData = countdata,
                              colData = coldata,
                              design = ~ project)

# pre-filtering to reduce memory size of dds, and to increase speed of count modeling 
smallestGroupSize <- 77
keep <- rowSums(counts(dds) >= 10) >= smallestGroupSize
dds <- dds[keep,]
```

## differential gene expression analysis with DESeq2

```{r deseq2}
dds_res <- DESeq(dds)
res <- results(dds_res)
summary(res)
res
```